solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -1588,12 +1589,39 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
       if (updateLastState) {
         cd.getCloudDescriptor().setLastPublished(state);
       }
-      overseerJobQueue.offer(Utils.toJSON(m));
+      DocCollection coll = zkStateReader.getCollection(collection);
+      if(forcePublish || sendToOverseer(coll, coreNodeName)) {
+        overseerJobQueue.offer(Utils.toJSON(m));
+      } else {
+        if(log.isDebugEnabled()) {
+          log.debug("bypassed overseer for message : {}", Utils.toJSONString(m));
+        }
+        PerReplicaStates perReplicaStates = coll.getPerReplicaStates(zkClient);
+        PerReplicaStates.WriteOps ops = PerReplicaStates.WriteOps.flipState(coreNodeName, state, perReplicaStates);
+        PerReplicaStates.persist(ops, perReplicaStates.path, zkClient);
+      }
     } finally {
       MDCLoggingContext.clear();
     }
   }
 
+  /**Whether a message needs to be sent to oversereer or not
+   */
+  static boolean sendToOverseer(DocCollection coll, String replicaName) {
+    if (coll == null) return true;
+    if (coll.getStateFormat() < 2 || !coll.isPerReplicaState()) return true;
+    Replica r = coll.getReplica(replicaName);
+    if (r == null) return true;
+    Slice shard = coll.getSlice(r.slice);
+    if (shard == null) return true;//very unlikely
```

Comment:
Unlikely bc this is a bad state, or how?

Reply:
yeah, it has to be a bad/invalid state